libobs/obs-nix-wayland.c

```diff
@@ -19,7 +19,224 @@
 #include "obs-nix-platform.h"
 #include "obs-nix-wayland.h"
 
+#include <unistd.h>   // close
+#include <sys/mman.h> // mmap
+#include <string.h>   // strcmp
+
 #include <wayland-client.h>
+#include <xkbcommon/xkbcommon.h>
+
+// X11 only supports 256 scancodes, most keyboards dont have 256 keys so this should be reasonable.
+#define MAX_KEYCODES 256
+// X11 keymaps only have 4 shift levels, im not sure xkbcommon supports a way to shift the state into a higher level anyway.
+#define MAX_SHIFT_LEVELS 4
+
+struct obs_hotkeys_platform {
+	struct wl_display *display;
+	struct wl_seat *seat;
+	struct wl_keyboard *keyboard;
+	struct xkb_context *xkb_context;
+	struct xkb_keymap *xkb_keymap;
+	struct xkb_state *xkb_state;
+	xkb_keysym_t key_to_sym[MAX_KEYCODES][MAX_SHIFT_LEVELS];
+	xkb_keysym_t obs_to_key[OBS_KEY_LAST_VALUE];
+};
+
+static obs_key_t obs_nix_wayland_key_from_virtual_key(int sym);
+
+static void load_keymap_data(struct xkb_keymap *keymap, xkb_keycode_t key,
+			     void *data)
+{
+	obs_hotkeys_platform_t *plat = (obs_hotkeys_platform_t *)data;
+	if (key >= MAX_KEYCODES)
+		return;
+
+	const xkb_keysym_t *syms;
+	// My setup doesnt have multiple layouts, do people actually use these?
+	// int num_layouts = xkb_keymap_num_layouts_for_key(keymap, key);
+	// blog(LOG_INFO, "key layouts: %d", num_layouts);
```

Comment:
based on https://www.x.org/releases/X11R7.6/doc/xorg-docs/input/XKB-Config.html#id2520758 and an overview of the keyboard layouts preloaded on archlinux. There are no multi-layout keyboards defined by default, the only case for this would be users with custom defined xkb keymaps (so like 3 people?) where there appears to be like 4 possible layout levels.

I dont mind also hardcoding in layouts if we assume there can be up to 4 of them, though im not sure we really want to support arbitrary groups too, especially not if IMEs do change keymaps instead.

Reply:
I played with some IME's and gnome, it looks like IMEs will use their own path for inputting characters for text input and do not use this mechanism at all.

For users configuring keyboards via gnome settings and not using IMEs gnome will only ever configure 4 layouts/groups at a time to remain X11 compatible. If you switch past the 4th keyboard gnome will reconfigure the wayland keyboard.

In my minimal playing this code seems to work fine for multiple languages once we handle multiple layouts/groups.

Reply:
Alright, thanks for checking this

Reply:
Pushed changes, I build the mapping tables any time users switch layouts. In cases where a user has something like a qwerty and azerty layout, scancode for AD01 might be bound to 'q' and 'a' which isnt quite correct. Switching keymaps takes 0.016ms on my machine so I think there is no problem with computing the lookup tables when users swap layouts to make the code simpler.